### PR TITLE
Allow deletion of Assignment's with no groups

### PR DESCRIPTION
**Pull request by mimischly7**
## Motivation and Context

An authorized instructor should be able to delete an assignment if and only if the assignment does not have any groups.
Although `routes.rb` already generated the endpoint for destroying an assignment, there was no `#destroy` action in `controllers/api/assignments_controller.rb`.  

button-enabled
button-disabled

## Your Changes

**Description**:
I implemented a `#destroy` action in `controllers/api/assignments_controller.rb` which deletes the Assignment corresponding to the given `id` if this assignment does not have any groups. The logic for checking whether an assignment has any groups is performed at the model level using the `dependent: restrict_with_exception` attribute of rails associations. The controller catches the potential exception and responds to the request accordingly.

**Type of change** (select all that apply):

- [x] New feature (non-breaking change which adds functionality)

## Testing

Wrote several rspec tests in `spec/controller/api/assignments_controller_spec.rb` aiming to capture the various scenarios and contexts under which one may attempt to delete an assignment.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the pre-commit.ci checks have passed. 
- [x] I have verified that the CI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have added tests for my changes. 
- [x] I have updated the Changelog.md file. 
- [x] I have made changes to the documentation and linked to that pull request below. 

### Pull request to make documentation changes (if applicable)

https://github.com/MarkUsProject/Wiki/pull/201

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
 - Bug Fix: Prevent assigning inactive graders to a group (#6801)
 - Added new API routes for the `create` and `destroy` actions of SectionsController (#6873)
 - Display error message in real-time when running automated tests, rather than waiting for page refresh (#6878)
+- - Added `destroy` action to `controllers/api/assignments_controller` (#6880)
```
Comment (david-yz-liu) [inline Changelog.md:10]: There's an extra `-` here. Also, please do a pull from master; we've made a new minor release, so you'll need to add this to the new `[unreleased]` section

**Review by david-yz-liu — COMMENTED** [review]
@mimischly7 good! So in addition to this change, I want you to add a button to delete the assignment *from the web interface* (and more specifically from the assignment edit form.

Add the button and give it a red background using one of the built-in colour constants, so that it's very clearly distinguished from other buttons. Make sure the button is disabled if there are any groups, and that there is a confirmation dialog box. This will need a parallel assignments controller route on the non-API controller.

**Review by david-yz-liu — COMMENTED** [review]
@mimischly7 great work! Sorry for the delay on reviewing this. I left a few more comments for you to polish this PR.

File: app/controllers/api/assignments_controller.rb [hunk 2]
```diff
@@ -299,6 +299,25 @@ def submit_file
       upload_file(grouping, only_required_files: assignment.only_required_files)
     end
 
+    def destroy
+      assignment_id = params[:id]
+      assignment = Assignment.find_by(id: assignment_id)
+      if assignment.nil?
+        render 'shared/http_status', locals: { code: '404', message:
+          I18n.t('assignments.assignment_not_found',
+                 invalid_id: assignment_id) }, status: :not_found
```
[the remark this answers is not in the document]
Comment (mimischly7) [reply app/controllers/api/assignments_controller.rb]: Given our discussion on Friday on whether we should check in the controller whether the requested record is actually valid, even though the `SessionHandler` class method `check_record` (as a `before_action` method) already checks that, I decided to keep the extra check. Let me know if you want me to remove it.

File: app/views/assignments/_form.html.erb [hunk 3]
```diff
@@ -489,4 +489,17 @@
                    data: { disable_with: t('working') } %>
     </p>
   <% end %>
+
+<!--Conditional Rendering-->
```
[the remark this answers is not in the document]
Comment (mimischly7) [reply app/views/assignments/_form.html.erb]: I applied this change and also used a ternary operator to account for some other differences (e.g. for the disabled case I wanted to let the button take default css classes).

File: app/views/assignments/_form.html.erb [hunk 4]
```diff
@@ -489,4 +489,17 @@
...
+    <%= button_to "Delete",
+                  course_assignment_path(@current_course, @assignment),
+                  data: { confirm: "Are you sure you want to delete this assignment?" },
+                  method: 'delete',
+                  class: 'delete-button',
+                  form_class: 'display-inline-block',
+                  title: t('helpers.submit.delete', model: Assignment.model_name.human),
+                  remote: true %>
+  <% end  %>
+
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/views/assignments/_form.html.erb:509]: This line hasn't been deleted

File: app/assets/stylesheets/common/_markus.scss [hunk 5]
```diff
@@ -303,6 +303,12 @@ button,
   }
 }
 
+#assignment-delete {
```
Comment (david-yz-liu) [inline app/assets/stylesheets/common/_markus.scss:306]: Okay this is better, but (1) make it a class, and (2) call the class `danger-button`, which is more general.

File: app/controllers/api/assignments_controller.rb [hunk 6]
```diff
@@ -299,6 +299,24 @@ def submit_file
       upload_file(grouping, only_required_files: assignment.only_required_files)
     end
 
+    def destroy
+      assignment = record
+      if assignment.nil?
```
Comment (david-yz-liu) [inline app/controllers/api/assignments_controller.rb:304]: This check isn't necessary, you can remove it. The `check_record` helper is sufficient to check for this

File: app/controllers/assignments_controller.rb [hunk 7]
```diff
@@ -657,6 +657,17 @@ def lti_settings
     render layout: 'assignment_content'
   end
 
+  def destroy
+    @assignment = @record
```
Comment (david-yz-liu) [inline app/controllers/assignments_controller.rb:661]: Use `record` (as with the other methods)

File: app/views/assignments/_form.html.erb [hunk 8]
```diff
@@ -489,4 +489,22 @@
                    data: { disable_with: t('working') } %>
     </p>
   <% end %>
+
+  <% if (action_name=="edit") %>
```
Comment (david-yz-liu) [inline app/views/assignments/_form.html.erb:493]: If statements don't need parentheses in Ruby. Also, put spaces around `==`

File: app/views/assignments/_form.html.erb [hunk 9]
```diff
@@ -489,4 +489,22 @@
...
+    <% has_groups = @assignment.groups.length != 0 %>
+    <%= button_to t(:delete),
+                  course_assignment_path(@current_course, @assignment),
+                  {
+                    data: { confirm: "Are you sure you want to delete this assignment?" },
```
Comment (david-yz-liu) [inline app/views/assignments/_form.html.erb:498]: This string also needs to be internationalized

File: app/views/assignments/_form.html.erb [hunk 10]
```diff
@@ -489,4 +489,22 @@
...
+                    method: 'delete',
+                    class: has_groups ? nil : 'button',
+                    id: has_groups ? nil : 'assignment-delete',
```
Comment (david-yz-liu) [inline app/views/assignments/_form.html.erb:501]: Following my above comment, don't use an `id` here, instead conditionally add a second class to `'button'` above

File: app/views/assignments/_form.html.erb [hunk 11]
```diff
@@ -489,4 +489,22 @@
...
+                    form_class: 'display-inline-block',
+                    title: has_groups ? I18n.t('assignments.assignment_has_groupings') :
+                             I18n.t('helpers.submit.delete', model: Assignment.model_name.human),
+                    disabled: has_groups
+                  }
+    %>
+  <%end %>
```
Comment (david-yz-liu) [inline app/views/assignments/_form.html.erb:508]: There is a space missing after the `<%`

File: spec/controllers/api/assignments_controller_spec.rb [hunk 12]
```diff
@@ -502,6 +507,35 @@
         expect(response).to have_http_status(403)
       end
     end
+    context 'DELETE assignment' do
+      it 'should successfully delete assignment because the assignment has no groups' do
+        assignment # since lazy let is used for creating an assignment, I invoke it here to trigger its execution
```
Comment (david-yz-liu) [inline spec/controllers/api/assignments_controller_spec.rb:512]: This isn't necessary, as `assignment` is immediately used on the next line

File: spec/controllers/api/assignments_controller_spec.rb [hunk 13]
```diff
@@ -502,6 +507,35 @@
...
+        expect(assignment.groups).to be_empty
+        delete :destroy, params: { id: assignment.id, course_id: course.id }
+        expect(response).to have_http_status(200)
+        expect(Assignment.exists?(assignment.id)).to eq(false)
+      end
+      it 'fails to delete assignment because assignment has groups' do
+        assignment # since lazy let is used for creating an assignment, I invoke it here to trigger its execution
```
Comment (david-yz-liu) [inline spec/controllers/api/assignments_controller_spec.rb:519]: Same comment as above

**Review by david-yz-liu — APPROVED** [review]

**Pull request merged**